rts/motoko-rts/src/gc.rs

```diff
@@ -1,2 +1,31 @@
 pub mod copying;
 pub mod mark_compact;
+
+#[cfg(feature = "ic")]
+unsafe fn should_do_gc() -> bool {
+    use crate::memory::ic::{HP, LAST_HP};
+
+    use core::cmp::{max, min};
+
+    // A factor of last heap size. We allow at most this much allocation before doing GC.
+    const HEAP_GROWING_FACTOR: f32 = 1.5;
+
+    // On small heaps `last_hp * HEAP_GROWING_FACTOR` will be quite small. To avoid doing redundant
+    // collections in such cases we only check `last_hp * HEAP_GROWING_FACTOR` if at least this
+    // much is allocated.
+    const SMALL_HEAP_DELTA: u32 = 10 * 1024 * 1024; // 10 MiB
+
+    // Regardless of other parameters (`HEAP_GROWING_FACTOR`, `SMALL_HEAP_DELTA`) we want to do GC
+    // if `HP` is more than this amount.
+    const MAX_HP_FOR_GC: u32 = 3 * 1024 * 1024 * 1024; // 3 GiB
```

Comment:
Should we start conservatively here? (e.g. <1GB) and increase it in subsequent releases?

Reply:
Doesn't this criterion mean that we'll GC on every message after we've reached MAX_HP_FOR_GC of live data? Shouldn't this still be taking the amount of allocation since last GC into account somehow (as a heuristic for amount of dead data)?

Reply:
My suggestion is about incremental rollout of the change: make the change for smaller heaps first, see that it works as expected, and then apply to larger ones. The motivation is that the risk of OOM is higher for larger heaps.

In general we do what a similar heap growing for large heap, but need to be more careful.

Reply:
@ulan, ah, sorry, my comment wasn't in reply to yours, just on the same line of code. :)  I agree with rolling out in stages. But I believe a fixed upper bound is gonna be a massive (prohibitive?) performance cliff?

Reply:
> Doesn't this criterion mean that we'll GC on every message after we've reached MAX_HP_FOR_GC of live data

Only as long as heap size is larger than `MAX_HP_FOR_GC`, but yes. So if after a message we free enough space we'll be back with `heap size * growth factor` check again.

> But I believe a fixed upper bound that forces back into per-message GC is gonna be a massive (prohibitive?) performance cliff?

This is possible, yeah. I guess we should try and see. The "massive" case is the default today so this should still be an improvement.

Motivations for `MAX_HP_FOR_GC` are:

- If we only use `heap size * growth factor` then a program that allocates little over many calls will fill the heap. So at some point we should do GC regardless of how much the last message allocated.
- The risk of OOM as mentioned by Ulan.

Reply:
> Should we start conservatively here? (e.g. <1GB) and increase it in subsequent releases?

Done. I'll update the plots.

Reply:
> I'll update the plots.

Done.